source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.md

```diff
@@ -592,6 +675,18 @@ try:
         if connection is perished:
           close connection
           connection = Null
+        if connection is in "pending response" state:
```

Comment:
Connection pools are FIFO.  Considering the case where we are successfully reading chunks of data off of the socket but we're timing out before we finish the read (the final else block in your pseudocode), won't checking the connection back into the pool and looping just cause this request to immediately checkout the same connection again (the behavior https://jira.mongodb.org/browse/DRIVERS-3276 is intending to avoid)? 

And regarding https://jira.mongodb.org/browse/DRIVERS-3276: are we convinced that its better to potentially block a number of operations on the slow connection than to block just one?

Reply:
We can check in the connection back only if the current check out is timed out (either because of CSOT or waitQueueTimeout has elapsed), but "pending response sliding window" has not expired yet. But the "current" operation is timed out anyway, so it should raise an error instead of attempting another connection.